media/multipart/src/main/java/io/helidon/media/multipart/MultiPartDecoder.java

```diff
@@ -106,120 +142,229 @@ public void onNext(DataChunk chunk) {
                 int id = parser.offer(byteBuffers[i]);
                 // record the chunk using the id of the last buffer
                 if (i == byteBuffers.length - 1) {
+                    // drain() cannot be invoked concurrently, it is safe to use HashMap
                     chunksByIds.put(id, chunk);
                 }
             }
-            parser.parse();
+            parserIterator = parser.parseIterator();
+            drain();
         } catch (MimeParser.ParsingException ex) {
-            emitter.fail(ex);
-            chunk.release();
-            releaseChunks();
-        }
-
-        // submit parsed parts
-        while (!bodyParts.isEmpty()) {
-            if (emitter.isCancelled()) {
-                return;
-            }
-            emitter.emit(bodyParts.poll());
-        }
-
-        // complete the parts publisher
-        if (parserEventProcessor.isCompleted()) {
-            emitter.complete();
-            // parts are delivered sequentially
-            // we potentially drop the last part if not requested
-            emitter.clearBuffer(this::drainPart);
-            releaseChunks();
-        }
-
-        // request more data to detect the next part
-        // if not in the middle of a part content
-        // or if the part content subscriber needs more
-        if (upstream != SubscriptionHelper.CANCELED
-                && emitter.hasRequests()
-                && parserEventProcessor.isDataRequired()
-                && (!parserEventProcessor.isContentDataRequired() || bodyPartPublisher.hasRequests())) {
-
-            upstream.request(1);
+            drain(ex);
         }
     }
 
     @Override
     public void onError(Throwable throwable) {
         Objects.requireNonNull(throwable);
-        initFuture.whenComplete((e, t) -> e.fail(throwable));
+        if (upstream != SubscriptionHelper.CANCELED) {
```

Comment:
Actually, this is what needs doing: remove the check and the following line setting `upstream = CANCELED;`. Then the subtle race condition discussed in the comment to `drainBoth` will disappear: then observing `upstream == CANCELED` is possible only if `onComplete` has been observed, which is mutually exclusive with `onError`.

Add `upstream = CANCELED;` to `cleanup`.

Reply:
Well, that won't work. The spec is very particular about not allowing synchronous cancellation, even though concurrent cancellation should be supported by `upstream`. And "they" will tell you "just add a check". It isn't about adding a check. It is about synchronizing on two variables instead of just one, for a dubious purpose.